opencog/atoms/distvalue/DistributionalValueSCM.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * opencog/atoms/distvalue/DistributionalValueSCM.h
+ *
+ * Copyright (c) 2018 by SingularityNet
+ * All Rights Reserved
+ *
+ * Written by Roman Treutlein
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef _OPENCOG_DISTRIBUTIONAL_VALUE_SCM_H
+#define _OPENCOG_DISTRIBUTIONAL_VALUE_SCM_H
+
+#include <opencog/atoms/distvalue/DistributionalValue.h>
+#include <opencog/atoms/distvalue/ConditionalDV.h>
+#include <opencog/atoms/distvalue/DVFormulas.h>
+
+#include <opencog/guile/SchemeModule.h>
+
+namespace opencog
+{
+
+class DistributionalValueSCM : public ModuleWrap
+{
+
+public:
+	DistributionalValueSCM();
+
+protected:
+	void init(void);
+
+	//C Value -> SCM
+	SCM float_list_to_scm(const std::vector<double>&);
+	SCM dvecseq_to_scm(const DVecSeq& v);
+	SCM dv_to_scm(const DistributionalValuePtr&);
+	SCM dvs_to_scm(const std::vector<DistributionalValuePtr>&);
+	SCM cdv_to_scm(const ConditionalDVPtr&);
+
+	//SCM -> C Value
+
+	DVecSeq verify_DVecSeq(SCM svalue_list, const char * subrname, int pos);
+	DVecSeq scm_to_DVecSeq(SCM svalue_list);
+	DistributionalValuePtr verify_dv(SCM, const char *, int pos = 1);
+	std::vector<DistributionalValuePtr> verify_dv_list(SCM, const char *,
+	                                                   int pos = 1);
+	ConditionalDVPtr verify_cdv(SCM, const char *, int pos = 1);
+
+	//SCHEME Functions
+	SCM ss_new_dv(SCM, SCM);
+	SCM ss_new_dv_simple(SCM, SCM);
+	SCM ss_dv_p(SCM);
+	SCM ss_dv_divide(SCM,SCM,SCM);
+	SCM ss_dv_sum_joint(SCM,SCM);
+	SCM ss_dv_get_confidence(SCM);
+	SCM ss_dv_conjunction(SCM,SCM);
+	SCM ss_dv_disjunction(SCM,SCM);
+	SCM ss_dv_negate(SCM);
+	SCM ss_dv_is_empty(SCM);
+	SCM ss_cdv_is_empty(SCM);
+	SCM ss_new_cdv(SCM, SCM);
+	SCM ss_cdv_get_confidence(SCM);
+	SCM ss_cdv_get_conditions(SCM);
+	SCM ss_cdv_get_unconditonals(SCM);
+	SCM ss_cdv_get_unconditonal(SCM,SCM);
+	SCM ss_cdv_get_joint(SCM,SCM);
+	SCM ss_cdv_merge(SCM,SCM);
+	SCM ss_cdv_join(SCM,SCM);
+
+	SCM ss_dv_merge_hi_conf(SCM,SCM);
+	SCM ss_cdv_merge_hi_conf(SCM,SCM);
+	//SCM ss_cdv_cde(SCM,SCM);
```

Comment:
Do you actually need all of the above? This is a pretty sharp and direct violation 
of the AtomSpace and scheme binding policy; the AtomSpace is not a programming
API meant for for human programmers!  Scheme (or python or haskell...) bindings
cannot be introspected, and thus are utterly useless and opaque for any algorithm.
Only human programmers can use the scheme bindings, and we are trying to 
keep humans out of this layer.  So .. again .. do you really need these?  Why?
Can you design it so that fewer of these (ideally, zero of these) are needed?

Reply:
I do need those as they are used by some PLN rules that work with DVs instead of TVs. Which i created to check weather this is actualy usable. But maybe this could be split off and moved to opencog.

Reply:
@rTreutlein FYI, I am urging @ngeiswei to convert the PLN formulas to use `PREDICATE_FORMULA_LINK` which evaluates the formulas using atomese (PlusLink, TimesLink, etc.) There are two reasons for this conversion:

1) It makes formulas at least 10x if not 100x faster than calling `GroundedPredicate "scm:foobar"` because calling `scm:foobar` incurs a 50 microsecond overhead (20K calls/sec) Same is true for `python:whatever`

2) In some whizzy future, things like `asmoses` or whatever learning algo could auto-learn new formulas, place them directly in the atomspace, and just go; no need for jumping through hoops to jam it into scm and then call that scm.

I figure you should know this, as it may alter your thinking about such things.

Reply:
See `examples/atomspace/formulas.scm` for a working example